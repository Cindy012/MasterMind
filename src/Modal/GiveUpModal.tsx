interface GiveUpModalProps {
    setShowModal: React.Dispatch<React.SetStateAction<boolean>>;
    hideCloseButton?: boolean;
    show: boolean;
    showAnswer: () => void;
    title: string;
}

const GiveUpModal:React.FC<GiveUpModalProps> = ({ setShowModal, show, hideCloseButton, showAnswer, title }) => {
    const stopGame = () => {
        showAnswer();
        setShowModal(false);
    };

    return (
        <div className={`modal ${show ? 'active' : ''}`}>
            <div className='modal__content'>
                { !hideCloseButton && <span onClick={() => setShowModal(false)} className="modal__close">&times;</span> }
                <h2>{ title }</h2> 
                <p>Are you sure to give up?</p>
                <div className="modal__footer">
                    <button className="modal__button" onClick={ () => stopGame() }>Yes</button>
                    <button className="modal__button" onClick={ () => setShowModal(false) }>No</button>
                    
                </div>
            </div>
        </div>
    );
};

export default GiveUpModal;
