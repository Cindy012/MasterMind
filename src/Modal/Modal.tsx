import confetti from 'canvas-confetti';
import { useEffect, useState } from 'react';

interface ModalProps {
    setShowModal: React.Dispatch<React.SetStateAction<boolean>>;
    hideCloseButton?: boolean;
    show: boolean;
    resetGame?: () => void
    gameStatus: number; // 0: gameInfo, 1: not filled/ongoing, 2: winner, 3: loser
    gameInfo?: JSX.Element;
}

const Modal:React.FC<ModalProps> = ({ setShowModal, show, hideCloseButton, resetGame, gameStatus, gameInfo }) => {
    const [playConfetti, setPlayConfetti] = useState(false);
    const playAgain = () => {
        if (resetGame) {
            setPlayConfetti(false);
            resetGame();
            setShowModal(false);
        }
    };

    const title: string[] = [
        'Mastermind game info',
        'Not Yet!',
        'Winner!',
        'Loser!'
    ];

    const text: string[] = [
        '',
        'You did not give the code colors.',
        'You win, play again!',
        'You lose, next time better!'
    ];

    if (gameStatus === 2 && resetGame && !playConfetti) {
        setPlayConfetti(true);
    }

    useEffect(() => {
        if (playConfetti) {
            confetti({
                particleCount: 160,
                spread: 100,
                colors: ['#e43b3b', '#e48a3b', '#e4de3b', '#3be45d', '#3be4de', '#3b60e4', '#8d3be4', '#e43b9b']
            });
        }
    },[playConfetti]);
       
    return (
        <div className={`modal ${show ? 'active' : ''}`}>
            <div className={ gameInfo ? 'modal__content' : 'modal__content modal__content__small' }>
                { !hideCloseButton && <span onClick={() => setShowModal(false)} className="modal__close">&times;</span> }
                <h2>{ title[gameStatus] }</h2> 
                { gameInfo && gameStatus === 0 ? gameInfo : <p>{ text[gameStatus] }</p> }
                <div className={ gameStatus === 0 || gameStatus === 1 ? 'modal__footer__secondary' : 'modal__footer'}>
                    { gameStatus !== 1 && gameStatus !== 0 ? (
                        <button className="modal__button" onClick={ () => playAgain() }>Play again!</button>
                    ) : null }
                    <button className="modal__button" onClick={ () => setShowModal(false) }>Close</button>
                </div>
            </div>
        </div>
    );
};

export default Modal;
