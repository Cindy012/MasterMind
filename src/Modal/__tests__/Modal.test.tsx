import React from 'react';
import '@testing-library/jest-dom';
import { render } from 'react-dom';
import Modal from '../Modal';
import GiveUpModal from '../GiveUpModal';

describe('Modal - situations:', () => {    
    it('renders the (GameInfo) Modal', () => {
        const gameStatus = 0;
        const setStateMocked = jest.fn();
        const useStateMock: any = (useState: any) => [useState, setStateMocked];
        jest.spyOn(React, 'useState').mockImplementation(useStateMock);
        const [showModalMock, setShowModalMock] = useStateMock(true);

        const container = document.createElement('div');
        render(<Modal setShowModal={ setShowModalMock } show={ showModalMock } gameStatus={ gameStatus } />, container);
        expect(container).toMatchSnapshot();
    });

    it('renders the (Not yet) Modal', () => {
        const gameStatus = 1;
        const setStateMocked = jest.fn();
        const useStateMock: any = (useState: any) => [useState, setStateMocked];
        jest.spyOn(React, 'useState').mockImplementation(useStateMock);
        const [showModalMock, setShowModalMock] = useStateMock(true);

        const container = document.createElement('div');
        render(<Modal setShowModal={ setShowModalMock } show={ showModalMock } gameStatus={ gameStatus } />, container);
        expect(container).toMatchSnapshot();
    });

    it('renders the (Winner) Modal', () => {
        const gameStatus = 2;
        const setStateMocked = jest.fn();
        const useStateMock: any = (useState: any) => [useState, setStateMocked];
        jest.spyOn(React, 'useState').mockImplementation(useStateMock);
        const [showModalMock, setShowModalMock] = useStateMock(true);

        const container = document.createElement('div');
        render(<Modal setShowModal={ setShowModalMock } show={ showModalMock } gameStatus={ gameStatus } />, container);
        expect(container).toMatchSnapshot();
    });

    it('renders the (Loser) Modal', () => {
        const gameStatus = 3;
        const setStateMocked = jest.fn();
        const useStateMock: any = (useState: any) => [useState, setStateMocked];
        jest.spyOn(React, 'useState').mockImplementation(useStateMock);
        const [showModalMock, setShowModalMock] = useStateMock(true);

        const container = document.createElement('div');
        render(<Modal setShowModal={ setShowModalMock } show={ showModalMock } gameStatus={ gameStatus } />, container);
        expect(container).toMatchSnapshot();
    });

    it('does not render the Modal successfully when gameStatus < gameStatus.length', () => {
        const gameStatus = -1;
        const setStateMocked = jest.fn();
        const useStateMock: any = (useState: any) => [useState, setStateMocked];
        jest.spyOn(React, 'useState').mockImplementation(useStateMock);
        const [showModalMock, setShowModalMock] = useStateMock(true);

        const container = document.createElement('div');
        render(<Modal setShowModal={ setShowModalMock } show={ showModalMock } gameStatus={ gameStatus } />, container);
        expect(container).toMatchSnapshot();
    });

    it('does not render the Modal successfully when gameStatus > gameStatus.length', () => {
        const gameStatus = 4;
        const setStateMocked = jest.fn();
        const useStateMock: any = (useState: any) => [useState, setStateMocked];
        jest.spyOn(React, 'useState').mockImplementation(useStateMock);
        const [showModalMock, setShowModalMock] = useStateMock(true);

        const container = document.createElement('div');
        render(<Modal setShowModal={ setShowModalMock } show={ showModalMock } gameStatus={ gameStatus } />, container);
        expect(container).toMatchSnapshot();
    });
});

describe('GiveUpModal', () => {
    it('renders the Modal', () => {
        const setStateMocked = jest.fn();
        const useStateMock: any = (useState: any) => [useState, setStateMocked];
        jest.spyOn(React, 'useState').mockImplementation(useStateMock);
        const [showGiveUpModalMock, setshowGiveUpModalMock] = useStateMock(true);

        const container = document.createElement('div');
        render(
            <GiveUpModal
                title='Stop Game'
                show={showGiveUpModalMock}
                setShowModal={setshowGiveUpModalMock}
                hideCloseButton 
                showAnswer={ function (): void {
                    throw new Error('Function not implemented.');
                }}
            />, container
        );
        expect(container).toMatchSnapshot();
    });
});