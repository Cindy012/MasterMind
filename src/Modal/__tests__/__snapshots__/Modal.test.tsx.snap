// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GiveUpModal renders the Modal 1`] = `
<div>
  <div
    class="modal active"
  >
    <div
      class="modal__content small"
    >
      <h2>
        Stop Game
      </h2>
      <p>
        Are you sure to give up?
      </p>
      <div
        class="modal__footer"
      >
        <button
          class="modal__button"
        >
          Yes
        </button>
        <button
          class="modal__button"
        >
          No
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal - situations: does not render the Modal successfully when gameStatus < gameStatus.length 1`] = `
<div>
  <div
    class="modal active"
  >
    <div
      class="modal__content small"
    >
      <span
        class="modal__close"
      >
        ×
      </span>
      <p>
        Oops, Something went wrong!
      </p>
      <div
        class="modal__footer"
      >
        <button
          class="modal__button"
        >
          Play again!
        </button>
        <button
          class="modal__button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal - situations: does not render the Modal successfully when gameStatus > gameStatus.length 1`] = `
<div>
  <div
    class="modal active"
  >
    <div
      class="modal__content small"
    >
      <span
        class="modal__close"
      >
        ×
      </span>
      <h2 />
      <p />
      <div
        class="modal__footer"
      >
        <button
          class="modal__button"
        >
          Play again!
        </button>
        <button
          class="modal__button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal - situations: renders the (GameInfo) Modal 1`] = `
<div>
  <div
    class="modal active"
  >
    <div
      class="modal__content null"
    >
      <span
        class="modal__close"
      >
        ×
      </span>
      <h2>
        Mastermind game info
      </h2>
      <article>
        <p>
          MasterMind is a game whereby the user guesses code on four pegs. The player has eight chances to guess the color code.
        </p>
        <p>
          To choose a color, select from the side one of the colors pegs. After selecting the color, the player can fill the row of the board.
        </p>
        <p>
          When the player selects a code pattern, the code will be reviewed and provides feedback through clues. The black peg means that the chosen peg color and position are correct. The red peg means that the peg color retains in the code although the position is not.
        </p>
      </article>
      <div
        class="modal__footer__secondary"
      >
        <button
          class="modal__button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal - situations: renders the (Loser) Modal 1`] = `
<div>
  <div
    class="modal active"
  >
    <div
      class="modal__content small"
    >
      <span
        class="modal__close"
      >
        ×
      </span>
      <h2>
        Loser!
      </h2>
      <p>
        You lose, next time better!
      </p>
      <div
        class="modal__footer"
      >
        <button
          class="modal__button"
        >
          Play again!
        </button>
        <button
          class="modal__button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal - situations: renders the (Not yet) Modal 1`] = `
<div>
  <div
    class="modal active"
  >
    <div
      class="modal__content small"
    >
      <span
        class="modal__close"
      >
        ×
      </span>
      <h2>
        Not Yet!
      </h2>
      <p>
        You did not give the code colors.
      </p>
      <div
        class="modal__footer__secondary"
      >
        <button
          class="modal__button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal - situations: renders the (Winner) Modal 1`] = `
<div>
  <div
    class="modal active"
  >
    <div
      class="modal__content small"
    >
      <span
        class="modal__close"
      >
        ×
      </span>
      <h2>
        Winner!
      </h2>
      <p>
        You win, play again!
      </p>
      <div
        class="modal__footer"
      >
        <button
          class="modal__button"
        >
          Play again!
        </button>
        <button
          class="modal__button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;
